llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp

```diff
@@ -885,55 +885,93 @@ template <typename Ty = llvm::Value> Ty *getVal(llvm::Metadata *M) {
   return nullptr;
 }
 
-/// Return the MDNode that has the SLM size attribute.
-static llvm::MDNode *getSLMSizeMDNode(llvm::Function *F) {
-  llvm::NamedMDNode *Nodes =
-      F->getParent()->getNamedMetadata(GENX_KERNEL_METADATA);
-  assert(Nodes && "invalid genx.kernels metadata");
-  for (auto Node : Nodes->operands()) {
-    if (Node->getNumOperands() >= 4 && getVal(Node->getOperand(0)) == F)
-      return Node;
-  }
-  // if F is not a kernel, keep looking into its callers
-  while (!F->use_empty()) {
-    auto CI = cast<CallInst>(F->use_begin()->getUser());
-    auto UF = CI->getParent()->getParent();
-    if (auto Node = getSLMSizeMDNode(UF))
-      return Node;
-  }
-  return nullptr;
-}
-
 static inline llvm::Metadata *getMD(llvm::Value *V) {
   return llvm::ValueAsMetadata::get(V);
 }
 
-static void translateSLMInit(CallInst &CI) {
-  auto F = CI.getParent()->getParent();
+/// Updates genx.kernels metadata attribute \p MD for the given function \p F.
+/// The value of the attribute is updated only if the new value \p NewVal is
+/// bigger than what is already stored in the attribute.
+// TODO: 1) In general this function is supposed to handle intrinsics
+// translated into kernel's metadata. So, the primary/intended usage model is
+// when such intrinsics are called from kernels.
+// 2) For now such intrinsics are also handled in functions directly called
+// from kernels and being translate into those caller-kernel meeven though such
+// behaviour is not fully specified/documented.
+// 3) This code (or the code in FE) must verify that slm_init or other such
+// intrinsic is not called from another module because kernels in that other
+// module would not get updated meta data attributes.
+static void updateGenXMDNodes(llvm::Function *F, genx::KernelMDOp MD,
+                              uint64_t NewVal) {
+  llvm::NamedMDNode *GenXKernelMD =
+      F->getParent()->getNamedMetadata(GENX_KERNEL_METADATA);
+  assert(GenXKernelMD && "invalid genx.kernels metadata");
 
-  auto *ArgV = CI.getArgOperand(0);
-  if (!isa<ConstantInt>(ArgV)) {
-    assert(false && "integral constant expected for slm size");
-    return;
-  }
-  auto NewVal = cast<llvm::ConstantInt>(ArgV)->getZExtValue();
-  assert(NewVal != 0 && "zero slm bytes being requested");
+  SmallPtrSet<Function *, 4> FunctionsVisited;
+  SmallVector<Function *, 4> FunctionsToVisit{F};
```

Comment:
I'd suggest to use more commonly adopted worklist pattern with `while (!empty)`, increase the initial size, as 4 might be frequently not enough, and not reuse F for clarity:

```cpp
SmallPtrSet<Function *, 32> FunctionsVisited;
SmallVector<Function *, 32> Worklist{F};

while (!Worklist.empty()) {
  Function *CurF = Worklist.pop_back_val();
  FunctionsVisited.insert(CurF);
  ...
    if (!FunctionsVisited.count(FCaller))
          Worklist.push_back(FCaller);
  ...
}
```

Reply:
Ok, Thank you. Fixed in ab21b2a33852d3b65ca8da1001c553733d77f1ae